File: frontend/src/shared/triggers.ts
import { createContentHashUuid } from './hashIds'

export type JenaTriggerId = string

export interface JenaTriggerMatcher {
  text: string
  isRegex: boolean
}

export interface JenaTimerEarlyEnder {
  text: string
  isRegex: boolean
}

export type JenaTriggerTimerType = 'countdown' | 'repeating' | 'stopwatch'

export type JenaTimerStartBehavior =
  | 'startNew'
  | 'restart'
  | 'restartMatchingTimerName'
  | 'ignoreIfRunning'

export interface JenaTrigger {
  id: JenaTriggerId
  name: string
  comments: string
  category: string
  groupPath: string[]
  match: JenaTriggerMatcher
  actions: JenaTriggerActions
  timer: JenaTriggerTimer | null
}

export interface JenaTriggerActions {
  display: JenaTextAction
  speech: JenaSpeechAction
  clipboard: JenaClipboardAction
}

export interface JenaTextAction {
  enabled: boolean
  text: string
}

export interface JenaSpeechAction {
  enabled: boolean
  text: string
  interrupt: boolean
}

export interface JenaClipboardAction {
  enabled: boolean
  text: string
}

export interface JenaTriggerTimer {
  type: JenaTriggerTimerType
  name: string
  durationMs: number
  startBehavior: JenaTimerStartBehavior
  warningSeconds: number
  warningAction: JenaTimerAction | null
  endedAction: JenaTimerAction | null
  earlyEnders: JenaTimerEarlyEnder[]
}

export interface JenaTimerAction {
  display: JenaTextAction
  speech: JenaSpeechAction
}

export interface JenaCharacterServer {
  characterName: string
  serverName: string
}

export interface JenaExtendedTrigger {
  triggerId: JenaTriggerId
  enabledFor: JenaCharacterServer[]
}

export interface JenaResolvedTrigger {
  trigger: JenaTrigger
  enabledFor: JenaCharacterServer[]
}

export interface JenaTriggerEnablementChange {
  triggerId: JenaTriggerId
  character: JenaCharacterServer
  enabled: boolean
}

export interface JenaTriggerUpsert {
  trigger: JenaTrigger
  enabledFor?: JenaCharacterServer[]
}

export interface JenaUserTriggerFetchResponse {
  records: JenaExtendedTrigger[]
  revision: string
  triggers: JenaTrigger[]
}

export interface JenaUserTriggerUpdate {
  deletedTriggerIds: JenaTriggerId[]
  revision: string
  upsertedRecords: JenaExtendedTrigger[]
  upsertedTriggers: JenaTrigger[]
}

export function createEmptyTrigger(): JenaTrigger {
  return {
    actions: {
      clipboard: createClipboardAction(),
      display: createTextAction(),
      speech: createSpeechAction(),
    },
    category: 'Default',
    comments: '',
    groupPath: [],
    id: 'draft-trigger',
    match: createTriggerMatcher(),
    name: '',
    timer: null,
  }
}

export function withCanonicalTriggerId(trigger: JenaTrigger): JenaTrigger {
  const content = getJenaTriggerHashContent(trigger)

  return {
    ...content,
    id: createContentHashUuid(content),
  }
}

export function createJenaTriggerId(trigger: JenaTrigger): JenaTriggerId {
  return createContentHashUuid(getJenaTriggerHashContent(trigger))
}

export function createTextAction(): JenaTextAction {
  return {
    enabled: false,
    text: '',
  }
}

export function createClipboardAction(): JenaClipboardAction {
  return {
    enabled: false,
    text: '',
  }
}

export function createSpeechAction(): JenaSpeechAction {
  return {
    enabled: false,
    interrupt: false,
    text: '',
  }
}

export function createTriggerMatcher(): JenaTriggerMatcher {
  return {
    text: '',
    isRegex: false,
  }
}

export function matcherToRegexSource(matcher: JenaTriggerMatcher) {
  return matcher.isRegex ? matcher.text : escapeRegExp(matcher.text)
}

function getJenaTriggerHashContent(trigger: JenaTrigger) {
  return {
    actions: getTriggerActionsHashContent(trigger.actions),
    category: trigger.category,
    comments: trigger.comments,
    groupPath: [...trigger.groupPath],
    match: getTriggerMatcherHashContent(trigger.match),
    name: trigger.name,
    timer: trigger.timer ? getTriggerTimerHashContent(trigger.timer) : null,
  }
}

function getTriggerMatcherHashContent(matcher: JenaTriggerMatcher) {
  return {
    text: matcher.text,
    isRegex: matcher.isRegex,
  }
}

function getTriggerActionsHashContent(actions: JenaTriggerActions) {
  return {
    display: getTextActionHashContent(actions.display),
    speech: getSpeechActionHashContent(actions.speech),
    clipboard: getTextActionHashContent(actions.clipboard),
  }
}

function getTriggerTimerHashContent(timer: JenaTriggerTimer) {
  return {
    type: timer.type,
    name: timer.name,
    durationMs: timer.durationMs,
    startBehavior: timer.startBehavior,
    warningSeconds: timer.warningSeconds,
    warningAction: timer.warningAction
      ? getTimerActionHashContent(timer.warningAction)
      : null,
    endedAction: timer.endedAction
      ? getTimerActionHashContent(timer.endedAction)
      : null,
    earlyEnders: timer.earlyEnders.map(getTriggerMatcherHashContent),
  }
}

function getTimerActionHashContent(action: JenaTimerAction) {
  return {
    display: getTextActionHashContent(action.display),
    speech: getSpeechActionHashContent(action.speech),
  }
}

function getTextActionHashContent(action: JenaTextAction | JenaClipboardAction) {
  return {
    enabled: action.enabled,
    text: action.text,
  }
}

function getSpeechActionHashContent(action: JenaSpeechAction) {
  return {
    enabled: action.enabled,
    text: action.text,
    interrupt: action.interrupt,
  }
}

function escapeRegExp(text: string) {
  return text.replace(/[.*+?^${}()|[\]\\]/g, '\\$&')
}

File: backend/model/triggers.go
package model

type TriggerID string

type TriggerMatcher struct {
	Text    string `json:"text"`
	IsRegex bool   `json:"isRegex"`
}

type TimerEarlyEnder struct {
	Text    string `json:"text"`
	IsRegex bool   `json:"isRegex"`
}

type TriggerTimerType string

const (
	TriggerTimerTypeCountdown TriggerTimerType = "countdown"
	TriggerTimerTypeRepeating TriggerTimerType = "repeating"
	TriggerTimerTypeStopwatch TriggerTimerType = "stopwatch"
)

type TimerStartBehavior string

const (
	TimerStartBehaviorStartNew                 TimerStartBehavior = "startNew"
	TimerStartBehaviorRestart                  TimerStartBehavior = "restart"
	TimerStartBehaviorRestartMatchingTimerName TimerStartBehavior = "restartMatchingTimerName"
	TimerStartBehaviorIgnoreIfRunning          TimerStartBehavior = "ignoreIfRunning"
)

type Trigger struct {
	ID        TriggerID      `json:"id"`
	Name      string         `json:"name"`
	Comments  string         `json:"comments"`
	Category  string         `json:"category"`
	GroupPath []string       `json:"groupPath"`
	Match     TriggerMatcher `json:"match"`
	Actions   TriggerActions `json:"actions"`
	Timer     *TriggerTimer  `json:"timer"`
}

type TriggerActions struct {
	Display   TextAction      `json:"display"`
	Speech    SpeechAction    `json:"speech"`
	Clipboard ClipboardAction `json:"clipboard"`
}

type TextAction struct {
	Enabled bool   `json:"enabled"`
	Text    string `json:"text"`
}

type SpeechAction struct {
	Enabled   bool   `json:"enabled"`
	Text      string `json:"text"`
	Interrupt bool   `json:"interrupt"`
}

type ClipboardAction struct {
	Enabled bool   `json:"enabled"`
	Text    string `json:"text"`
}

type TriggerTimer struct {
	Type           TriggerTimerType   `json:"type"`
	Name           string             `json:"name"`
	DurationMs     int64              `json:"durationMs"`
	StartBehavior  TimerStartBehavior `json:"startBehavior"`
	WarningSeconds int64              `json:"warningSeconds"`
	WarningAction  *TimerAction       `json:"warningAction"`
	EndedAction    *TimerAction       `json:"endedAction"`
	EarlyEnders    []TimerEarlyEnder  `json:"earlyEnders"`
}

type TimerAction struct {
	Display TextAction   `json:"display"`
	Speech  SpeechAction `json:"speech"`
}

type CharacterServer struct {
	CharacterName string `json:"characterName"`
	ServerName    string `json:"serverName"`
}

type ExtendedTrigger struct {
	TriggerID  TriggerID         `json:"triggerId"`
	EnabledFor []CharacterServer `json:"enabledFor"`
	Publish    bool              `json:"publish"`
	Broadcast  bool              `json:"broadcast"`
}

type ResolvedTrigger struct {
	Trigger    Trigger           `json:"trigger"`
	EnabledFor []CharacterServer `json:"enabledFor"`
	Publish    bool              `json:"publish"`
	Broadcast  bool              `json:"broadcast"`
}

type TriggerEnablementChange struct {
	TriggerID TriggerID       `json:"triggerId"`
	Character CharacterServer `json:"character"`
	Enabled   bool            `json:"enabled"`
}

type TriggerFlagChange struct {
	TriggerID TriggerID `json:"triggerId"`
	Publish   *bool     `json:"publish,omitempty"`
	Broadcast *bool     `json:"broadcast,omitempty"`
}

type TriggerUpsert struct {
	Trigger    Trigger           `json:"trigger"`
	EnabledFor []CharacterServer `json:"enabledFor,omitempty"`
}

type UserTriggerUpdate struct {
	DeletedTriggerIDs []TriggerID       `json:"deletedTriggerIds"`
	Revision          string            `json:"revision"`
	UpsertedRecords   []ExtendedTrigger `json:"upsertedRecords"`
	UpsertedTriggers  []Trigger         `json:"upsertedTriggers"`
}
